Add a few peepholes

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62914 Add handling of ifs to shape propagation
* #62911 Small shape analysis changes
* **#62910 Add a few peepholes**
* #61975 Propagate symbolic dimensions through idioms like x.view(y.size())

Differential Revision: [D30196947](https://our.internmc.facebook.com/intern/diff/D30196947)

Commit:
Add a few peepholes

[ghstack-poisoned]

Gamrix commented on torch/csrc/jit/passes/peephole_non_tensor.cpp:
```diff
@@ -192,25 +193,49 @@ struct PeepholeOptimizeNonTensorImpl {
-            break;
+        auto immut_type = [&](const TypePtr& type) {
+          auto kind = type->kind();
+          std::vector<TypeKind> handled_immutable_types = {
```
This should be able to have a `static const` header, right?

Gamrix commented on torch/csrc/jit/passes/peephole_non_tensor.cpp:
```diff
@@ -192,25 +193,49 @@ struct PeepholeOptimizeNonTensorImpl {
+              TypeKind::IntType,
+              TypeKind::FloatType,
+              TypeKind::NoneType};
+          return (
+              std::find(
+                  handled_immutable_types.begin(),
+                  handled_immutable_types.end(),
+                  kind) != handled_immutable_types.end());
```
This would be a more readable statement if it was broken into two parts.

Gamrix commented on torch/csrc/jit/passes/peephole_non_tensor.cpp:
```diff
@@ -192,25 +193,49 @@ struct PeepholeOptimizeNonTensorImpl {
           node->output()->replaceAllUsesWith(node->input());
           changed = true;
         }
+      } else if (node->kind() == aten::Int) {
+        if (node->input()->type()->cast<IntType>()) {
+          GRAPH_UPDATE(
+              "Removing ", getHeader(node), " as input is already an integer");
+          node->output()->replaceAllUsesWith(node->input());
+          changed = true;
+        }
```
Why is this not handled by the code right above it?

eellison replied:
aten::Int is converting scalar/float/bool to int. so shouldnt be handled by unchecked_cast

Gamrix commented on test/jit/test_peephole.py:
```diff
@@ -506,13 +506,18 @@ def foo(x: int, y: int):
+            return x == x, x != x
+
+        inps = 1, [2, 3]
+        for func, inp in zip([foo, foo2], inps):
```
rewriting the zip -> [[foo, 1], [foo2, [2, 3]] honestly would be shorter and clearer.

Gamrix reviewed (approved):
Added comments on some small nits. Overall looks good to go.

Commit:
Update on "Add a few peepholes"




[ghstack-poisoned]

Commit:
Update on "Add a few peepholes"




[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30196947).

Commit:
Update on "Add a few peepholes"


Differential Revision: [D30196947](https://our.internmc.facebook.com/intern/diff/D30196947)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30196947).

Commit:
Update on "Add a few peepholes"


Differential Revision: [D30196947](https://our.internmc.facebook.com/intern/diff/D30196947)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30196947).
